### PR TITLE
[SPARK-25387][SQL] Fix for NPE caused by bad CSV input

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

The PR fixes NPE in `UnivocityParser` caused by malformed CSV input. In some cases, `uniVocity` parser can return `null` for bad input. In the PR, I propose to check result of parsing and not propagate NPE to upper layers.

## How was this patch tested?

I added a test which reproduce the issue and tested by `CSVSuite`.

### CODE REVIEW

**Comment by maropu** [comment]
This line below possibly returns null?
https://github.com/apache/spark/blob/77c996403d5c761f0dfea64c5b1cb7480ba1d3ac/sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L510

**Comment by maropu** [comment]
ok, thanks for the check!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 1]
```diff
@@ -1700,4 +1700,13 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
     checkCount(2)
     countForMalformedCSV(0, Seq(""))
   }
+
+  test("SPARK-25387: bad input should not cause NPE") {
+    val schema = StructType(StructField("a", IntegerType) :: Nil)
+    val input = spark.createDataset(Seq("\u0000\u0000\u0001234"))
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: btw, in this title, bad CSV means what (bad unicode?)? In this case, the CSV parser returns null and, in another case, it throws `com.univocity.parsers.common.TextParsingException`? I just want to know the behaivour in the parser.

Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: The `parseLine` method can return `null` in many cases. See:
https://github.com/uniVocity/univocity-parsers/blob/f616d151b48150bc9cb98943f9b6f8353b704359/src/main/java/com/univocity/parsers/common/AbstractParser.java#L663
https://github.com/uniVocity/univocity-parsers/blob/f616d151b48150bc9cb98943f9b6f8353b704359/src/main/java/com/univocity/parsers/common/AbstractParser.java#L678

It is normal way for the method to indicate about an error.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala [hunk 2]
```diff
@@ -243,17 +243,21 @@ object TextInputCSVDataSource extends CSVDataSource {
...
-      val caseSensitive = sparkSession.sessionState.conf.caseSensitiveAnalysis
-      val header = makeSafeHeader(firstRow, caseSensitive, parsedOptions)
-      val sampled: Dataset[String] = CSVUtils.sample(csv, parsedOptions)
-      val tokenRDD = sampled.rdd.mapPartitions { iter =>
-        val filteredLines = CSVUtils.filterCommentAndEmpty(iter, parsedOptions)
-        val linesWithoutHeader =
-          CSVUtils.filterHeaderLine(filteredLines, firstLine, parsedOptions)
-        val parser = new CsvParser(parsedOptions.asParserSettings)
-        linesWithoutHeader.map(parser.parseLine)
+      if (firstRow != null) {
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: Can we simplify the code as 
```
    maybeFirstLine.map(new CsvParser(parsedOptions.asParserSettings).parseLine(_)) match {
      case Some(firstRow) if firstRow != null =>
      case _ =>
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala [hunk 3]
```diff
@@ -216,7 +216,12 @@ class UnivocityParser(
   }
 
   private def convert(tokens: Array[String]): InternalRow = {
-    if (tokens.length != parsedSchema.length) {
+    if (tokens == null) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: when will we hit this?

Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: I got it on a CSV files with some marks at the beginning but `uniVocity` parser returns `null` in many cases when it cannot read/parse input, for example: https://github.com/uniVocity/univocity-parsers/blob/f616d151b48150bc9cb98943f9b6f8353b704359/src/main/java/com/univocity/parsers/common/AbstractParser.java#L663

**Comment by cloud-fan** [comment]
thanks, merging to master/2.4!

**Review by HyukjinKwon — COMMENTED** [review]
LGTM too

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala [hunk 4]
```diff
@@ -240,23 +240,25 @@ object TextInputCSVDataSource extends CSVDataSource {
...
+        val header = makeSafeHeader(firstRow, caseSensitive, parsedOptions)
+        val sampled: Dataset[String] = CSVUtils.sample(csv, parsedOptions)
+        val tokenRDD = sampled.rdd.mapPartitions { iter =>
+          val filteredLines = CSVUtils.filterCommentAndEmpty(iter, parsedOptions)
+          val linesWithoutHeader =
+            CSVUtils.filterHeaderLine(filteredLines, maybeFirstLine.get, parsedOptions)
+          val parser = new CsvParser(parsedOptions.asParserSettings)
+          linesWithoutHeader.map(parser.parseLine)
+        }
+        CSVInferSchema.infer(tokenRDD, header, parsedOptions)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: @MaxGekk, BTW what happen if the second line is the malfromed record and it returns `null`? From a cursory look, schema inference looks going to throw an NPE exception.

Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: @HyukjinKwon I have checked this on (with header too):
```Scala
    val input = spark.createDataset(Seq("1", "\u0000\u0000\u0001234"))

    val df = spark.read.option("inferSchema", true).csv(input)
    df.printSchema()
    df.show()
```
```
root
 |-- _c0: integer (nullable = true)

+----+
| _c0|
+----+
|   1|
|null|
+----+
```
In the debugger, I didn't observe `null` in https://github.com/apache/spark/blob/5264164a67df498b73facae207eda12ee133be7d/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala#L61-L69 .

**Pull request closed without merging**